src/build/calibrate/calibrate_national.jl

```diff
@@ -0,0 +1,183 @@
+    io::Dict,
+    set::Dict;
+    save_build::Bool=SLiDE.DEFAULT_SAVE_BUILD,
+    overwrite::Bool=SLiDE.DEFAULT_OVERWRITE,
+    penalty_nokey::AbstractFloat=SLiDE.DEFAULT_PENALTY_NOKEY,
+)
+    subset = "calibrate"
+
+    # If there is already calibration data, read it and return.
+    d_read = SLiDE.read_build(dataset, subset; overwrite=overwrite)
+    !(isempty(d_read)) && (return d_read)
+    
+    # Initialize a DataFrame to contain results and do the calibration iteratively.
+    cal = Dict(k => DataFrame() for k in list_parameters!(set, :calibrate))
+
+    for year in set[:yr]
+        cal_yr = calibrate_national(io, set, year; penalty_nokey=penalty_nokey)
+        [cal[k] = [cal[k]; cal_yr[k]] for k in keys(cal_yr)]
+    end
+    
+    # If no DataFrame was returned by the annual calibrations, replace this with the input.
+    [cal[k] = io[k] for (k,df) in cal if isempty(df)]
+
+    SLiDE.write_build!(dataset, subset, cal; save_build=save_build)
+    return cal
+end
+
+
+function calibrate_national(
+    io::Dict,
+    set::Dict,
+    year::Int;
+    penalty_nokey::AbstractFloat=SLiDE.DEFAULT_PENALTY_NOKEY,
+    # multipliers for lower and upper bound relative
+    # to each respective variables reference parameter
+    lower_bound = SLiDE.DEFAULT_CALIBRATE_LOWER_BOUND,
+    upper_bound = SLiDE.DEFAULT_CALIBRATE_UPPER_BOUND,
+)
+    @info("Calibrating $year data")
+
+    # Save relevant sets to simplify things a little.
+    S, G, M, VA, FD = set[:s], set[:g], set[:m], set[:va], set[:fd]
+    
+    SLiDE._calibration_set!(set; final_demand=true, value_added=true)
+    d = SLiDE._national_calibration_input(io, set, year;
+        lower_bound=lower_bound,
+        upper_bound=upper_bound,
+    )
+    
+    calib = Model(optimizer_with_attributes(Ipopt.Optimizer, "max_cpu_time"=>60.0))
+    
+    @variables(calib, begin
+        fd0[g in G, fd in FD], (start=d[:fd0][g,fd], lower_bound=d[:fd0_lb][g,fd], upper_bound=d[:fd0_ub][g,fd])
+        va0[va in VA, s in S], (start=d[:va0][va,s], lower_bound=d[:va0_lb][va,s], upper_bound=d[:va0_ub][va,s])
+        id0[g in G, s in S], (start=d[:id0][g,s], lower_bound=d[:id0_lb][g,s], upper_bound=d[:id0_ub][g,s])
+        ys0[s in S, g in G], (start=d[:ys0][s,g], lower_bound=d[:ys0_lb][s,g], upper_bound=d[:ys0_ub][s,g])
+        md0[m in M, g in G], (start=d[:md0][m,g], lower_bound=d[:md0_lb][m,g], upper_bound=d[:md0_ub][m,g])
+        ms0[g in G, m in M], (start=d[:ms0][g,m], lower_bound=d[:ms0_lb][g,m], upper_bound=d[:ms0_ub][g,m])
+        a0[g in G],  (start=d[:a0][g], lower_bound=d[:a0_lb][g],  upper_bound=d[:a0_ub][g])
+        m0[g in G],  (start=d[:m0][g], lower_bound=d[:m0_lb][g],  upper_bound=d[:m0_ub][g])
+        x0[g in G],  (start=d[:x0][g], lower_bound=d[:x0_lb][g],  upper_bound=d[:x0_ub][g])
+        y0[g in G],  (start=d[:y0][g], lower_bound=d[:y0_lb][g],  upper_bound=d[:y0_ub][g])
+        fs0[g in G], (start=d[:fs0][g],lower_bound=d[:fs0_lb][g], upper_bound=d[:fs0_ub][g])
+    end)
```

Comment:
All variables defined in a block; `start`, `lower_bound`, and `upper_bound` set upon definition